File: bin/fnio.py
#
# Copyright (C) 2017-2020 Dimitar Toshkov Zhekov <dimitar.zhekov@gmail.com>
#
# This program is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by the Free
# Software Foundation; either version 2 of the License, or (at your option)
# any later version.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
# or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
# for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
#

import codecs
import struct
import sys
import os

# -- InputFileStream --
class InputFileStream:
	def __init__(self, file_name, encoding='binary'):
		if file_name is not None:
			self.file = open(file_name, 'r') if encoding is None else open(file_name, 'rb')
			self.st_name = file_name
		else:
			self.file = sys.stdin if encoding is None else sys.stdin.buffer
			self.st_name = '<stdin>'

		if encoding not in [None, 'binary']:
			self.file = codecs.getreader(encoding)(self.file)

		self.line_no = 0
		self.eof = False


	def close(self):
		self.unseek()
		self.file.close()


	def fstat(self):
		return None if (self.file == sys.stdin.buffer or self.file.isatty()) else os.fstat(self.file.fileno())


	def location(self):
		return '%s:%s' % (self.st_name, 'EOF: ' if self.eof else '%d: ' % self.line_no if self.line_no > 0 else ' ')


	def process(self, callback):
		try:
			result = callback(self)
			self.close()
			return result
		except Exception as ex:
			ex.message = self.location() + getattr(ex, 'message', str(ex))
			raise


	def read_line(self):
		return self.read_lines(lambda line: line)


	def read_lines(self, callback):
		try:
			for line in self.file:
				self.line_no += 1
				self.eof = False
				line = callback(line.rstrip())
				if line is not None:
					return line
		except OSError:
			self.unseek()
			raise

		self.eof = True
		return None


	def unseek(self):
		self.line_no = 0
		self.eof = False


# -- OutputFileStream --
class OutputFileStream:
	def __init__(self, file_name, encoding='binary'):
		if file_name is not None:
			self.file = open(file_name, 'wb')
			self.st_name = file_name
		else:
			self.file = sys.stdout.buffer
			self.st_name = '<stdout>'

		if encoding is None and self.file.isatty():
			raise Exception(self.location() + 'binary output may not be send to a terminal')

		self.encoding = (None if encoding == 'binary' else encoding)
		self.close_attempt = False


	def abort(self):
		errors = ''

		if self.file != sys.stdout.buffer:
			if not self.close_attempt:
				try:
					self.close()
				except Exception as ex:
					errors += '\n%sclose: %s' % (self.location(), str(ex))

			try:
				os.remove(self.st_name)
			except Exception as ex:
				errors += '\n%sunlink: %s' % (self.location(), str(ex))

		return errors


	def close(self):
		self.close_attempt = True
		self.file.close()


	def location(self):
		return self.st_name + ': '


	def process(self, callback):
		try:
			callback(self)
			self.close()
		except Exception as ex:
			ex.message = self.location() + getattr(ex, 'message', str(ex)) + self.abort()
			raise


	def write(self, data):
		self.file.write(data)


	def write8(self, value):
		self.write(struct.pack('B', value))


	def write16(self, value):
		self.write(struct.pack('<H', value))


	def write32(self, value):
		self.write(struct.pack('<L', value))


	def write_line(self, text):
		self.write((text if self.encoding is None else bytes(text, self.encoding)) + b'\n')


	def write_prop(self, name, value):
		self.write_line((bytes(name, 'ascii') + b' ' + value).rstrip())


	def write_zstr(self, bstr, num_zeros):
		self.write(bstr + bytes(num_zeros))


# -- read/write file --
def read_file(file_name, callback, encoding='binary'):
	return InputFileStream(file_name, encoding).process(callback)


def write_file(file_name, callback, encoding='binary'):
	return OutputFileStream(file_name, encoding).process(callback)

File: bin/fnio.js
/*
  Copyright (C) 2017-2020 Dimitar Toshkov Zhekov <dimitar.zhekov@gmail.com>

  This program is free software; you can redistribute it and/or modify it
  under the terms of the GNU General Public License as published by the Free
  Software Foundation; either version 2 of the License, or (at your option)
  any later version.

  This program is distributed in the hope that it will be useful, but
  WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
  or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
  for more details.

  You should have received a copy of the GNU General Public License along
  with this program; if not, write to the Free Software Foundation, Inc.,
  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
*/

'use strict';

const tty = require('tty');
const fs = require('fs');

// -- InputFileStream --
const BLOCK_SIZE = 4096;

class InputFileStream {
	constructor(fileName, encoding = 'binary') {
		if (fileName != null) {
			this.fd = fs.openSync(fileName, 'r');
			this.stName = fileName;
		} else {
			this.fd = process.stdin.fd;
			this.stName = '<stdin>';
		}
		this.encoding = encoding;
		this.unseek();
		this.lines = [];
		this.index = 0;
		this.buffer = Buffer.alloc(BLOCK_SIZE);
		this.remainder = '';
	}

	close() {
		this.unseek();
		fs.closeSync(this.fd);
	}

	fstat() {
		return (this.fd === process.stdin.fd || tty.isatty(this.fd)) ? null : fs.fstatSync(this.fd);
	}

	location() {
		let location = ' ';

		if (this.eof) {
			location = 'EOF: ';
		} else if (this.lineNo > 0) {
			location = `${this.lineNo}: `;

		}
		return `${this.stName}:${location}`;
	}

	_readBlock() {
		for (;;) {
			try {
				return fs.readSync(this.fd, this.buffer, 0, BLOCK_SIZE);
			} catch (e) {
				if (e.code === 'EOF') {
					return 0;
				}
				if (e.code !== 'EAGAIN') {
					this.unseek();
					throw e;
				}
			}
		}
	}

	readLine() {
		return this.readLines(line => line);
	}

	readLines(callback) {
		let line;

		do {
			while (this.index < this.lines.length) {
				this.lineNo++;
				line = callback(this.lines[this.index++].trimRight());

				if (line != null) {
					return line;
				}
			}

			var count = this._readBlock();

			this.index = 0;
			this.lines = (this.remainder + this.buffer.toString(this.encoding, 0, count)).split('\n');
			this.remainder = this.lines.pop();
			this.eof = false;
		} while (count > 0);

		if (this.remainder.length > 0) {
			this.lineNo++;
			line = callback(this.remainder.trimRight());
			this.remainder = '';
		} else {
			this.eof = true;
			line = null;
		}

		return line;
	}

	unseek() {
		this.lineNo = 0;
		this.eof = false;
	}
}

// -- OutputFileStream --
class OutputFileStream {
	constructor(fileName, encoding = 'binary') {
		if (fileName != null) {
			this.fd = fs.openSync(fileName, 'w');
			this.stName = fileName;
		} else {
			this.fd = process.stdout.fd;
			this.stName = '<stdout>';
		}
		if (encoding == null && tty.isatty(this.fd)) {
			throw new Error(this.location() + 'binary output may not be send to a terminal');
		}
		this.encoding = (encoding == null ? 'binary' : encoding);
		this.fbbuf = Buffer.alloc(4);
		this.closeAttempt = false;
	}

	close() {
		this.closeAttempt = true;
		fs.closeSync(this.fd);
	}

	destroy() {
		let errors = '';

		if (this.fd !== process.stdout.fd) {
			if (!this.closeAttempt) {
				try {
					fs.closeSync(this.fd);
				} catch (e) {
					errors += `\n${this.stName}: close: ${e.message}`;
				}
			}

			try {
				fs.unlinkSync(this.stName);
			} catch (e) {
				errors += `\n${this.stName}: unlink: ${e.message}`;
			}
		}

		return errors;
	}

	location() {
		return this.stName + ': ';
	}

	write(buffer) {
		fs.writeSync(this.fd, buffer, 0, buffer.length);
	}

	write8(value) {
		this.fbbuf.writeUInt8(value, 0);
		fs.writeSync(this.fd, this.fbbuf, 0, 1);
	}

	write16(value) {
		this.fbbuf.writeUInt16LE(value, 0);
		fs.writeSync(this.fd, this.fbbuf, 0, 2);
	}

	write32(value) {
		this.fbbuf.writeUInt32LE(value, 0);
		fs.writeSync(this.fd, this.fbbuf, 0, 4);
	}

	writeLine(text) {
		fs.writeSync(this.fd, text + '\n', null, this.encoding);
	}

	writeProp(name, value) {
		this.writeLine((name + ' ' + value).trimRight());
	}

	writeZStr(bstr, numZeros) {
		fs.writeSync(this.fd, bstr, null, 'binary');
		this.write(Buffer.alloc(numZeros));
	}
}

// -- Export --
module.exports = Object.freeze({
	InputFileStream,
	OutputFileStream
});
